src/Data/Text.hs

```diff
@@ -1623,7 +1533,9 @@ group = groupBy (==)
 -- | /O(n)/ Return all initial segments of the given 'Text', shortest
 -- first.
 inits :: Text -> [Text]
-inits t@(Text arr off len) = loop 0
+inits t@(Text arr off len)
+    | null t = [empty]
+    | otherwise = empty : loop (iter_ t 0)
```

Comment:
You can be even lazier.

```
inits t@(Text arr off len) = empty : loop 0
```

Reply:
Thanks, good spot! I tried to do something like this before but got myself confused with some edge cases. I think I've got it right in e275071d13230027004ad2047205c01e27be1bba

Reply:
Ah I see the issue is that `iter_` is implemented in terms of indexing, and to get the loop invariant right, we want to let `iter_` give us an index > len, but that can't happen. I think what I had before while not the nicest is the best we can do, but I might be wrong

Reply:
It would be a shame to make it less lazy than it needs (which would also require delaying the `Text arr off len` pattern BTW). That should be possible by making the test `i >= len` *before* calling `iter_`.

Reply:
I've moved the guard into the tail cell in e27e5fd8ff3fafcf74b415fc0748ade07bf2415c. Is something like that what you had in mind?

Reply:
I'll also see about making the pattern lazier

Reply:
I made it lazier in 92b8cfc7b29a929abff0542e9bc00914804674db.
Now the text is only scrutinized when forcing the tail

Reply:
<details>
  <summary>Core for inits</summary>


```
24447   │ inits
24448   │   = \ t_a69c ->
24449   │       : empty
24450   │         (case t_a69c of wild_X1 { Text bx_dah5 bx1_dah6 bx2_dah7 ->
24451   │          case <=# bx2_dah7 0# of {
24452   │            __DEFAULT ->
24453   │              case indexWord8Array# bx_dah5 bx1_dah6 of r#_ibhc { __DEFAULT ->
24454   │              let {
24455   │                c#_sbr3
24456   │                  = word2Int# (clz8# (and# (not# (word8ToWord# r#_ibhc)) 255##)) } in
24457   │              let { lvl185_sbTS = : wild_X1 [] } in
24458   │              letrec {
24459   │                $wloop_sdFw
24460   │                  = \ ww_sdFt ->
24461   │                      case >=# ww_sdFt bx2_dah7 of {
24462   │                        __DEFAULT ->
24463   │                          : (Text bx_dah5 bx1_dah6 ww_sdFt)
24464   │                            (case indexWord8Array# bx_dah5 (+# bx1_dah6 ww_sdFt) of r#1_X2
24465   │                             { __DEFAULT ->
24466   │                             let {
24467   │                               c#1_sbr5
24468   │                                 = word2Int# (clz8# (and# (not# (word8ToWord# r#1_X2)) 255##)) } in
24469   │                             $wloop_sdFw (+# ww_sdFt (xorI# c#1_sbr5 (<=# c#1_sbr5 0#)))
24470   │                             });
24471   │                        1# -> lvl185_sbTS
24472   │                      }; } in
24473   │              $wloop_sdFw (xorI# c#_sbr3 (<=# c#_sbr3 0#))
24474   │              };
24475   │            1# -> []
24476   │          }
24477   │          })
```

</section>